### PR TITLE
handle `HttpException`s coming from `ChromeTab.connect`

### LINKED ISSUE #153972: `HttpException: Connection closed before full header was received, uri = http://localhost:52097/json`
Found on 3.24.1 (12.1% of all reports as of today, 8/22/24).

There is no stack trace available (possibly due to https://github.com/dart-lang/sdk/issues/56474 or similar).

The source is probably https://github.com/google/webkit_inspection_protocol.dart/blob/82f0c1c46dfdba5edf7c5fa84456233121dd69e1/lib/webkit_inspection_protocol.dart#L53, which the flutter tool reaches through a few places including

https://github.com/flutter/flutter/blob/b023fad9e5583cc4c0f69ebff2243451f3def460/packages/flutter_tools/lib/src/isolated/resident_web_runner.dart#L615-L617

and 

https://github.com/flutter/flutter/blob/cb28dec614e481c5ebde52b545118202200c5d4c/packages/flutter_tools/lib/src/web/chrome.dart#L479-L480

and

https://github.com/flutter/flutter/blob/cb28dec614e481c5ebde52b545118202200c5d4c/packages/flutter_tools/lib/src/web/chrome.dart#L511-L517
.

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes https://github.com/flutter/flutter/issues/153972  (unless the cause of https://github.com/flutter/flutter/issues/153064#issuecomment-2305662791 happens to also prevent this fix from working).

In this PR, I've looked for all non-test call sites of `ChromeConnection.getTabs` and made sure are all wrapped in `try` blocks that handle `IOException` (`HttpException` is what we see in crash reporting, but I figure any `IOException` might as well be the same for all intents and purposes).

I plan on cherry-picking this the stable branch.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart [hunk 1]
```diff
@@ -608,9 +608,15 @@ Please provide a valid TCP port (an integer between 0 and 65535, inclusive).
...
-      }, retryFor: const Duration(seconds: 5));
+      late final ChromeTab? chromeTab;
+      try {
+        chromeTab = await chrome.chromeConnection.getTab((ChromeTab chromeTab) {
+          return !chromeTab.url.startsWith('chrome-extension');
+        }, retryFor: const Duration(seconds: 5));
+      } on IOException {
+        // We were unable to unable to communicate with Chrome.
+        chromeTab = null;
+      }
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:619]: This should already be caught by the recently added

https://github.com/flutter/flutter/blob/8b3cd9a8d4db6baa03bf1ae8326254fed2b4e942/packages/flutter_tools/lib/src/isolated/resident_web_runner.dart#L386-L389

However, I prefer trying to catch exceptions as close to the thrower as we can.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:619]: The semantics are different whether this is caught at line 386 or 616. That is to say:

1. should we be calling `appFailedToStart()` here?
2. should we be printing to STDERR the error and stacktrace?
3. should we be calling the more explicit `HttpException` here?
4. on the other hand, should we be calling the more general `IOException` on line 386?

I'm genuinely asking, I don't know this code at all.

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 2]
```diff
@@ -490,6 +490,10 @@ class Chromium {
         if (i == attempts) {
           rethrow;
         }
+      } on IOException catch (_) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/chrome.dart:493]: this looks weird, is the analyzer mad if this line is just: `} on IOException {` without the catch?

Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/chrome.dart:493]:
```suggestion
      } on IOException {
```

I made a suggestion, so we can smoke test the CLABot thing

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart:493]: Did you? I still see the CLA check as green (the bot shouldn't break until I accept the suggestion, if at all).

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart:493]: Oh I misread this. Sure lets try it

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart]: Looks like it still passed :)

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart]: addressed in [don't use catch(_)](https://github.com/flutter/flutter/pull/153978/commits/3f12eba91996e497918a61be55444e43c83b4005)

**Review by christopherfujino — APPROVED** [review]
This is beyond me, but at a high level it *looks* like the kind of thing that would catch this :)

**Pull request merged**